### PR TITLE
feat: add downscoping with Credential Access Boundaries samples

### PR DESCRIPTION
**Pull request by bojeil-google**
Adds a sample to illustrate token broker and token consumer functionality.

This includes samples on:

- How to initialize the CAB rules.
- How to initialize the downscoped credentials.
- How to inject the downscoped credentials in the token consumer.

The samples also include a test that verifies readonly access to a temporary
object.

### CODE REVIEW

**Comment by bojeil-google** [comment]
Regarding the checklist:
- I have followed the sample guidelines.
- A README was added.
- Tests passed when I ran them locally.
- Lint passed.
- No new API needs to be enabled (it should already be enabled).
- No new environment variables are needed (reuses existing ones).
- Please merge the PR for me once it is approved.
- This new sample is added to the Auth samples directory which already has code owners.

**Comment by bojeil-google** [comment]
@busunkim96, @arithmetic1728 FYI

File: auth/downscoping/snippets.py
Comment (busunkim96) [inline auth/downscoping/snippets.py:3]:
```suggestion
# Copyright 2021 Google LLC
```

File: auth/downscoping/snippets_test.py
Comment (busunkim96) [inline auth/downscoping/snippets_test.py:3]:
```suggestion
# Copyright 2021 Google LLC
```

File: auth/downscoping/snippets.py
Comment (bojeil-google) [reply auth/downscoping/snippets.py]: Thanks. Done.

File: auth/downscoping/snippets_test.py
Comment (bojeil-google) [reply auth/downscoping/snippets_test.py]: Thanks. Done.

File: auth/downscoping/snippets_test.py
Comment (busunkim96) [inline auth/downscoping/snippets_test.py:26]: Another option is to create and tear down the bucket and blob in the test.

See this example: https://github.com/GoogleCloudPlatform/python-docs-samples/blob/33e3de913c30d430acc040094264b36142c04d89/storage/s3-sdk/s3_sdk_test.py#L47-L65

File: auth/downscoping/snippets.py
Comment (busunkim96) [inline auth/downscoping/snippets.py:160]: Note: samples do not have to show how to setup/teardown all the resources they need.  As an example, [this sample](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/33e3de913c30d430acc040094264b36142c04d89/storage/s3-sdk/list_gcs_objects.py) is focused on showing how to list objects in a blob, so it _doesn't_ show how to create a storage bucket and add objects to it.

If having the sample be runnable via a `main()` is by design please disregard. :)

File: auth/downscoping/snippets.py
Comment (busunkim96) [inline auth/downscoping/snippets.py:48]: Is this going to be presented in the page that explains how to construct these availability expressions? If not, please consider adding a link to the docs that explain how these work.

**Comment by busunkim96** [comment]
@bojeil-google I took a first pass! If you have a TW you're working with I would recommend also getting them to review the snippets.

File: auth/downscoping/snippets_test.py
Comment (bojeil-google) [reply auth/downscoping/snippets_test.py]: Done.

File: auth/downscoping/snippets.py
Comment (bojeil-google) [reply auth/downscoping/snippets.py]: I moved the setup to the test.

File: auth/downscoping/snippets.py
Comment (bojeil-google) [reply auth/downscoping/snippets.py:51]: The plan is to present this in the docs page. I have added more comments to help explain each step.

File: auth/downscoping/snippets.py
Comment (bojeil-google) [reply auth/downscoping/snippets.py]: Done

File: auth/downscoping/snippets.py
Comment (bojeil-google) [reply auth/downscoping/snippets.py:31]: Thanks. Done.

**Review by bojeil-google — COMMENTED** [review]
Thanks for the review! All the comments should be addressed.

**Review by busunkim96 — APPROVED** [review]

File: auth/downscoping/snippets.py
Comment (arithmetic1728) [inline auth/downscoping/snippets.py:113]: why `object_name[0:3]`?

File: auth/downscoping/README.md
Comment (arithmetic1728) [inline auth/downscoping/README.md:14]: Can compute engine credentials be the source credentials?

File: auth/downscoping/snippets.py
Comment (bojeil-google) [reply auth/downscoping/snippets.py:113]: We want to test the ability to limit access to objects with a certain prefix string in the resource bucket:
https://cloud.google.com/iam/docs/downscoping-short-lived-credentials#example-multiple-buckets

This is an arbitrary index I chose. I didn't want to hardcode the string.

File: auth/downscoping/README.md
Comment (bojeil-google) [reply auth/downscoping/README.md:14]: Yeah they should work. These are basically service account tokens and they can be downscoped.

File: auth/downscoping/snippets.py
Comment (arithmetic1728) [reply auth/downscoping/snippets.py:113]: maybe adding a comment here to say `object_name[0:3]` is used as an object prefix

**Review by arithmetic1728 — APPROVED** [review]

File: auth/downscoping/snippets.py
Comment (busunkim96) [reply auth/downscoping/snippets.py:113]: To remove room for confusion, maybe pass it as a named argument `object_prefix=object_name[0:3]` or instantiate a `object_prefix` variable separately?

```py
# Use the first three characters of the object as the prefix.
object_prefix = object_name[0:3] 
return get_token_from_broker(bucket_name, object_prefix)
```

File: auth/downscoping/snippets.py
Comment (hegemonic) [inline auth/downscoping/snippets.py:106]: I'm a little fuzzy on what this means. We say this is "optional," but I think what we really mean is "we're not passing in the token/expiry in this example, because for demonstration purposes, we generate the token locally." True?

Also, should this say "...will generate a new downscoped token..."?

File: auth/downscoping/snippets.py
Comment (hegemonic) [inline auth/downscoping/snippets.py:121]: s/storage/Cloud Storage/

**Review by hegemonic — CHANGES_REQUESTED** [review]

File: auth/downscoping/README.md
Comment (bojeil-google) [reply auth/downscoping/README.md]: Done

File: auth/downscoping/snippets.py
Comment (bojeil-google) [reply auth/downscoping/snippets.py]: Done

File: auth/downscoping/snippets.py
Comment (bojeil-google) [reply auth/downscoping/snippets.py]: Done

File: auth/downscoping/snippets.py
Comment (bojeil-google) [reply auth/downscoping/snippets.py]: Done

File: auth/downscoping/snippets.py
Comment (bojeil-google) [reply auth/downscoping/snippets.py]: done

File: auth/downscoping/snippets.py
Comment (bojeil-google) [reply auth/downscoping/snippets.py]: Done

File: auth/downscoping/snippets.py
Comment (bojeil-google) [reply auth/downscoping/snippets.py]: Done

File: auth/downscoping/snippets.py
Comment (bojeil-google) [reply auth/downscoping/snippets.py]: Done

File: auth/downscoping/snippets.py
Comment (bojeil-google) [reply auth/downscoping/snippets.py]: I agree this is confusing. I have rephrased to explain why we are passing a `refresh_handler`. I think this is better.

File: auth/downscoping/snippets.py
Comment (bojeil-google) [reply auth/downscoping/snippets.py]: I added a comment explaining this and I am now passing a named argument `object_prefix=object_name[0:3]`

File: auth/downscoping/README.md
Comment (bojeil-google) [reply auth/downscoping/README.md]: Done

File: auth/downscoping/snippets.py
Comment (bojeil-google) [reply auth/downscoping/snippets.py]: I rephrased to: allow the consumer to obtain new downscoped tokens...
I think this is better as this snippet is part of the consumer code.

**Pull request merged**
